# Capstone Project: Determination of Toronto Neighborhood that Best Suits Mr. Smith Preferences

#### Step 1: Install necessary packages for the analysis (Geopy and Folium)

In [1]:
!conda install -c conda-forge geopy --yes
!conda install -c conda-forge folium=0.5.0 --yes

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2020.4.5.1 |       hecc5488_0         146 KB  conda-forge
    certifi-2020.4.5.1         |   py36h9f0ad1d_0         151 KB  conda-forge
    geopy-1.21.0               |             py_0          58 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    openssl-1.1.1f             |       h516909a_0         2.1 MB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0         conda-forge
    geopy:           1

#### Step 2: Import necessary packages for analysis (i.e. pandas, numpy, etc.)

In [3]:
import pandas as pd
import numpy as np
import io
import requests
from sklearn.cluster import KMeans
import folium
import matplotlib.cm as cm
import matplotlib.colors as colors
import json
from geopy.geocoders import Nominatim
import requests
from pandas.io.json import json_normalize

#### Step 3: Import lists of Postal codes, Borough and Neighborhoods in Toronto from Wikipedia. All postal codes beginning with M are located within the city of Toronto. 

In [4]:
df = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')[0]

#### Step 4: Data cleaning performed. 
1. Postal codes with no Boroughs assigned are removed from the list. This means that neighborhood does not exist in those areas. 
2. For postal codes with assigned boroughs but not assigned neighborhoods, the neighborhood name is assumed to be the borough name. 

In [5]:
df['Borough'].replace("Not assigned", np.NaN, inplace=True)
df.dropna(subset=['Borough'], axis=0, inplace=True)
df["Neighborhood"] = df["Neighborhood"].str.replace(" /",",")
df["Neighborhood"].replace("Not assigned", df["Borough"], inplace=True)
df.reset_index(drop=True, inplace=True)
df.head()

Postal code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [6]:
df.shape

(103, 3)

After data cleaning, 103 unique postal codes (i.e. neighborhoods) exist in Toronto City. 

#### Step 5: Extract data from the csv sheet containing the geospatial coordinates of each postcode. Join both tables together based on the postal codes to create a new dataframe. 

In [7]:
url="http://cocl.us/Geospatial_data"
s=requests.get(url).content
data_lat_log=pd.read_csv(io.StringIO(s.decode('utf-8')))
df_new = df.merge(data_lat_log, left_on='Postal code', right_on="Postal Code")
df_new.drop('Postal Code', 1, inplace=True)
df_new.head()

Postal code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [15]:
df_new["Borough"].value_counts()

North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
York                 5
East York            5
East Toronto         5
Mississauga          1
Name: Borough, dtype: int64

In [16]:
toronto_data=df_new;

#### Step 6: Determine Toronto, Canada Coordinates for Folium Map

In [17]:
address = 'Toronto, ON, Canada'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto, ON, Canada are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto, ON, Canada are 43.6534817, -79.3839347.


#### Step 7: Show Toronto on Map and show neighborhoods on the map. 

In [20]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Step 8: Foursquare API Client ID and SECRET for data extraction 

In [26]:
CLIENT_ID = '0NEISBFG0NMNCFEV3VSB453GQ2OWCD4EOXWCSUBZ0XW3I1AU' # your Foursquare ID
CLIENT_SECRET = 'LZJEMZVSDGVN434JTJ0MJZV2TDO3AYMRCBDSVMONVIC5REH1' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 0NEISBFG0NMNCFEV3VSB453GQ2OWCD4EOXWCSUBZ0XW3I1AU
CLIENT_SECRET:LZJEMZVSDGVN434JTJ0MJZV2TDO3AYMRCBDSVMONVIC5REH1


#### Step 9: Get the nearby venues for each neighborhood centre (based on coordinates) on Foursquare. Radius from the coordinates is set as 1000 m and limit results to 100. 

In [29]:
def getNearbyVenues(names, latitudes, longitudes, radius=1000, LIMIT=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [30]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  );

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview
The Danforth West, Ri

In [33]:
print(toronto_venues.shape)
toronto_venues.head()

(4952, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Parkwoods              43.753259              -79.329656   
1    Parkwoods              43.753259              -79.329656   
2    Parkwoods              43.753259              -79.329656   
3    Parkwoods              43.753259              -79.329656   
4    Parkwoods              43.753259              -79.329656   

               Venue  Venue Latitude  Venue Longitude        Venue Category  
0    Allwyn's Bakery       43.759840       -79.324719  Caribbean Restaurant  
1    Brookbanks Park       43.751976       -79.332140                  Park  
2        Tim Hortons       43.760668       -79.326368                  Café  
3                A&W       43.760643       -79.326865  Fast Food Restaurant  
4  Bruno's valu-mart       43.746143       -79.324630         Grocery Store

In [34]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                              50   
Alderwood, Long Branch                                                 25   
Bathurst Manor, Wilson Heights, Downsview North                        31   
Bayview Village                                                        15   
Bedford Park, Lawrence Manor East                                      41   
Berczy Park                                                           100   
Birch Cliff, Cliffside West                                            12   
Brockton, Parkdale Village, Exhibition Place                          100   
Business reply mail Processing CentrE                                  48   
CN Tower, King and Spadina, Railway Lands, Harb...                     15   
Caledonia-Fairbanks                                                    23   
Canada Post Gateway Processing Centre                                  54   
Cedarbrae                                                              31   
Central Bay Street                                                    100   
Christie                                                              100   
Church and Wellesley                                                  100   
Clarks Corners, Tam O'Shanter, Sullivan                                40   
Cliffside, Cliffcrest, Scarborough Village West                        12   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                            100   
Davisville North                                                      100   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                       18   
Don Mills                                                              76   
Dorset Park, Wexford Heights, Scarborough Town ...                     45   
Downsview                                                              71   
Dufferin, Dovercourt Village                                           68   
East Toronto                                                           95   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                     16   
Fairview, Henry Farm, Oriole                                           44   
First Canadian Place, Underground city                                100   
...                                                                   ...   
Roselawn                                                               21   
Rouge Hill, Port Union, Highland Creek                                  5   
Runnymede, Swansea                                                     81   
Runnymede, The Junction North                                          48   
Scarborough Village                                                    13   
South Steeles, Silverstone, Humbergate, Jamesto...                     20   
St. James Town                                                        100   
St. James Town, Cabbagetown                                            34   
Steeles West, L'Amoreaux West                                          30   
Stn A PO Boxes                                                        100   
Studio District                                                       100   
Summerhill West, Rathnelly, South Hill, Forest ...                     80   
The Annex, North Midtown, Yorkville                                   100   
The Beaches                                                            76   
The Danforth West, Riverdale                                          100   
The Kingsway, Montgomery Road, Old Mill North                          47   
Thorncliffe Park                                                       49   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                        100   
Victoria Village                                

#### Step 10: Check how many unique venue categories are there in each neighborhood. 

In [35]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 327 uniques categories.


#### Step 11: Apply one hot encoding to transform venue categories to categorical variables. 

In [40]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Zoo  Accessories Store  Afghan Restaurant  Airport  American Restaurant  \
0    0                  0                  0        0                    0   
1    0                  0                  0        0                    0   
2    0                  0                  0        0                    0   
3    0                  0                  0        0                    0   
4    0                  0                  0        0                    0   

   Amphitheater  Animal Shelter  Antique Shop  Aquarium  Arcade  ...  \
0             0               0             0         0       0  ...   
1             0               0             0         0       0  ...   
2             0               0             0         0       0  ...   
3             0               0             0         0       0  ...   
4             0               0             0         0       0  ...   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                 0            0                      0                0   
1                 0            0                      0                0   
2                 0            0                      0                0   
3                 0            0                      0                0   
4                 0            0                      0                0   

   Whisky Bar  Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0           0         0          0            0              0            0  
1           0         0          0            0              0            0  
2           0         0          0            0              0            0  
3           0         0          0            0              0            0  
4           0         0          0            0              0            0  

[5 rows x 327 columns]

In [37]:
toronto_onehot.shape

(4952, 327)

#### Step 12: Determine the mean frequency of each venue category in the neighborhood. 

In [41]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Zoo  Accessories Store  \
0                                           Agincourt  0.0           0.000000   
1                              Alderwood, Long Branch  0.0           0.000000   
2     Bathurst Manor, Wilson Heights, Downsview North  0.0           0.000000   
3                                     Bayview Village  0.0           0.000000   
4                   Bedford Park, Lawrence Manor East  0.0           0.000000   
5                                         Berczy Park  0.0           0.000000   
6                         Birch Cliff, Cliffside West  0.0           0.000000   
7        Brockton, Parkdale Village, Exhibition Place  0.0           0.010000   
8               Business reply mail Processing CentrE  0.0           0.000000   
9   CN Tower, King and Spadina, Railway Lands, Har...  0.0           0.000000   
10                                Caledonia-Fairbanks  0.0           0.000000   
11              Canada Post Gateway Processing Centre  0.0           0.000000   
12                                          Cedarbrae  0.0           0.000000   
13                                 Central Bay Street  0.0           0.000000   
14                                           Christie  0.0           0.000000   
15                               Church and Wellesley  0.0           0.000000   
16            Clarks Corners, Tam O'Shanter, Sullivan  0.0           0.000000   
17    Cliffside, Cliffcrest, Scarborough Village West  0.0           0.000000   
18                     Commerce Court, Victoria Hotel  0.0           0.000000   
19                                         Davisville  0.0           0.000000   
20                                   Davisville North  0.0           0.000000   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn  0.0           0.000000   
22                                          Don Mills  0.0           0.000000   
23  Dorset Park, Wexford Heights, Scarborough Town...  0.0           0.022222   
24                                          Downsview  0.0           0.000000   
25                       Dufferin, Dovercourt Village  0.0           0.000000   
26                                       East Toronto  0.0           0.000000   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...  0.0           0.000000   
28                       Fairview, Henry Farm, Oriole  0.0           0.000000   
29             First Canadian Place, Underground city  0.0           0.000000   
..                                                ...  ...                ...   
67                                           Roselawn  0.0           0.000000   
68             Rouge Hill, Port Union, Highland Creek  0.0           0.000000   
69                                 Runnymede, Swansea  0.0           0.000000   
70                      Runnymede, The Junction North  0.0           0.000000   
71                                Scarborough Village  0.0           0.000000   
72  South Steeles, Silverstone, Humbergate, Jamest...  0.0           0.000000   
73                                     St. James Town  0.0           0.000000   
74                        St. James Town, Cabbagetown  0.0           0.000000   
75                      Steeles West, L'Amoreaux West  0.0           0.000000   
76                                     Stn A PO Boxes  0.0           0.000000   
77                                    Studio District  0.0           0.000000   
78  Summerhill West, Rathnelly, South Hill, Forest...  0.0           0.000000   
79                The Annex, North Midtown, Yorkville  0.0           0.000000   
80                                        The Beaches  0.0           0.000000   
81                       The Danforth West, Riverdale  0.0           0.000000   
82      The Kingsway, Montgomery Road, Old Mill North  0.0           0.000000   
83                                   Thorncliffe Park  0.0           0.000000   
84           Toronto Dominion Centre, Design Exchange  0.0           0.0

In [42]:
toronto_grouped.shape

(97, 327)

#### Step 13: Find the top 5 (most common) venues in each neighborhood. 

In [44]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [80]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted_most = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_most['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted_most.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_most.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt    Chinese Restaurant   
1                           Alderwood, Long Branch        Discount Store   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village         Grocery Store   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Shopping Mall                Bakery           Coffee Shop   
1              Pharmacy           Pizza Place          Skating Rink   
2           Pizza Place     Convenience Store           Coffee Shop   
3                  Bank           Gas Station   Japanese Restaurant   
4           Coffee Shop            Restaurant  Fast Food Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Sandwich Place           Pizza Place  Caribbean Restaurant   
1        Sandwich Place         Garden Center           Gas Station   
2                 Trail        Sandwich Place               Dog Run   
3          Intersection    Chinese Restaurant                  Park   
4        Sandwich Place                  Bank             Juice Bar   

   8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0  Sri Lankan Restaurant     Indian Restaurant             Discount Store  
1             Donut Shop          Liquor Store              Shopping Mall  
2    Fried Chicken Joint    Frozen Yogurt Shop  Middle Eastern Restaurant  
3          Shopping Mall            Restaurant                       Café  
4        Thai Restaurant            Bagel Shop                     Bakery

#### Step 14: Find the bottom 5 (least common) venues in each neighborhood. 

In [81]:
def return_least_common_venues(row, num_bottom_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=True)
    
    return row_categories_sorted.index.values[0:num_bottom_venues]

In [84]:
num_bottom_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of least common venues
columns = ['Neighborhood']
for ind in np.arange(num_bottom_venues):
    try:
        columns.append('{}{} Least Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Least Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted_least = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_least['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted_least.iloc[ind, 1:] = return_least_common_venues(toronto_grouped.iloc[ind, :], num_bottom_venues)

neighborhoods_venues_sorted_least.head()

Neighborhood 1st Least Common Venue  \
0                                        Agincourt                    Zoo   
1                           Alderwood, Long Branch                    Zoo   
2  Bathurst Manor, Wilson Heights, Downsview North                    Zoo   
3                                  Bayview Village                    Zoo   
4                Bedford Park, Lawrence Manor East                    Zoo   

  2nd Least Common Venue   3rd Least Common Venue 4th Least Common Venue  \
0            Music Store             Music School                 Museum   
1              Nightclub  New American Restaurant             Nail Salon   
2             Nail Salon              Music Venue            Music Store   
3           Optical Shop              Opera House                 Office   
4             Nail Salon              Music Venue            Music Store   

  5th Least Common Venue 6th Least Common Venue   7th Least Common Venue  \
0          Moving Target          Movie Theater      Moroccan Restaurant   
1            Music Venue            Music Store             Music School   
2           Music School                 Museum            Moving Target   
3           Noodle House              Nightclub  New American Restaurant   
4           Music School                 Museum            Moving Target   

  8th Least Common Venue 9th Least Common Venue 10th Least Common Venue  
0            Music Venue    Monument / Landmark       Mobile Phone Shop  
1                 Museum          Moving Target           Movie Theater  
2          Movie Theater    Moroccan Restaurant     Monument / Landmark  
3             Nail Salon        Organic Grocery             Music Venue  
4          Movie Theater    Moroccan Restaurant     Monument / Landmark

#### Step 15: Join the top (most common) 10 venue categories and bottom (least common) 10 venue categories together to create a new dataframe consisting of both most common and least common data. 

In [190]:
data_toronto_filtered = neighborhoods_venues_sorted_most.merge(neighborhoods_venues_sorted_least, left_on='Neighborhood', right_on="Neighborhood")
#df_new.drop('Postal Code', 1, inplace=True)
data_toronto_filtered.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt    Chinese Restaurant   
1                           Alderwood, Long Branch        Discount Store   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village         Grocery Store   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Shopping Mall                Bakery           Coffee Shop   
1              Pharmacy           Pizza Place          Skating Rink   
2           Pizza Place     Convenience Store           Coffee Shop   
3                  Bank           Gas Station   Japanese Restaurant   
4           Coffee Shop            Restaurant  Fast Food Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Sandwich Place           Pizza Place  Caribbean Restaurant   
1        Sandwich Place         Garden Center           Gas Station   
2                 Trail        Sandwich Place               Dog Run   
3          Intersection    Chinese Restaurant                  Park   
4        Sandwich Place                  Bank             Juice Bar   

   8th Most Common Venue 9th Most Common Venue  ... 1st Least Common Venue  \
0  Sri Lankan Restaurant     Indian Restaurant  ...                    Zoo   
1             Donut Shop          Liquor Store  ...                    Zoo   
2    Fried Chicken Joint    Frozen Yogurt Shop  ...                    Zoo   
3          Shopping Mall            Restaurant  ...                    Zoo   
4        Thai Restaurant            Bagel Shop  ...                    Zoo   

  2nd Least Common Venue   3rd Least Common Venue 4th Least Common Venue  \
0            Music Store             Music School                 Museum   
1              Nightclub  New American Restaurant             Nail Salon   
2             Nail Salon              Music Venue            Music Store   
3           Optical Shop              Opera House                 Office   
4             Nail Salon              Music Venue            Music Store   

  5th Least Common Venue 6th Least Common Venue   7th Least Common Venue  \
0          Moving Target          Movie Theater      Moroccan Restaurant   
1            Music Venue            Music Store             Music School   
2           Music School                 Museum            Moving Target   
3           Noodle House              Nightclub  New American Restaurant   
4           Music School                 Museum            Moving Target   

  8th Least Common Venue 9th Least Common Venue 10th Least Common Venue  
0            Music Venue    Monument / Landmark       Mobile Phone Shop  
1                 Museum          Moving Target           Movie Theater  
2          Movie Theater    Moroccan Restaurant     Monument / Landmark  
3             Nail Salon        Organic Grocery             Music Venue  
4          Movie Theater    Moroccan Restaurant     Monument / Landmark  

[5 rows x 21 columns]

In [104]:
data_toronto_filtered.shape

(97, 21)

#### Step 16: Run K-means clustering algorithm for the entire dataset. 

In [189]:
# set number of clusters
from sklearn.cluster import KMeans
kclusters = 7
toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)
# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
# kmeans.labels_[0:20] 

In [191]:
# add clustering labels
data_toronto_filtered.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(data_toronto_filtered.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             3.0                  Park   
1  43.725882 -79.315572             3.0           Coffee Shop   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             0.0  Fast Food Restaurant   
4  43.662301 -79.389494             1.0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  ...  \
0     Convenience Store              Pharmacy         Shopping Mall  ...   
1            Boxing Gym          Hockey Arena   Sporting Goods Shop  ...   
2               Theater                 Diner                  Café  ...   
3           Coffee Shop            Restaurant     Accessories Store  ...   
4                  Park      Sushi Restaurant    Italian Restaurant  ...   

  1st Least Common Venue   2nd Least Common Venue 3rd Least Common Venue  \
0                    Zoo  New American Restaurant             Nail Salon   
1                    Zoo             Optical Shop            Opera House   
2                    Zoo              Music Venue            Music Store   
3                    Zoo               Nail Salon            Music Venue   
4                    Zoo              Music Venue            Music Store   

  4th Least Common Venue 5th Least Common Venue 6th Least Common Venue  \
0            Music Venue            Music Store           Music School   
1                 Office           Noodle House              Nightclub   
2           Music School          Moving Target          Movie Theater   
3            Music Store           Music School                 Museum   
4          Moving Target          Movie Theater    Moroccan Restaurant   

    7th Least Common Venue 8th Least Common Venue 9th Least Common Venue  \
0                   Museum          Moving Target          Movie Theater   
1  New American Restaurant             Nail Salon        Organic Grocery   
2      Moroccan Restaurant             Nail Salon    Monument / Landmark   
3            Moving Target          Movie Theater    Moroccan Restaurant   
4      Monument / Landmark             Nail Salon      Mobile Phone Shop   

  10th Least Common Venue  
0     Moroccan Restaurant  
1             Music Venue  
2       Mobile Phone Shop  
3     Monument / Landmark  
4           Metro Station  

[5 rows x 26 columns]

#### Step 17: Let's have a look on the results cluster by cluster.

# Cluster 1 (below) are neighborhoods most common for restaurants / pubs / coffee shops. It is definitely not somewhere where Mr. Smith wants to live in.

In [247]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
3      North York             0.0       Fast Food Restaurant   
6     Scarborough             0.0       Fast Food Restaurant   
7      North York             0.0                 Restaurant   
10     North York             0.0              Grocery Store   
13     North York             0.0                 Restaurant   
16           York             0.0          Convenience Store   
18    Scarborough             0.0                Pizza Place   
23      East York             0.0                Coffee Shop   
29      East York             0.0                Coffee Shop   
32    Scarborough             0.0             Ice Cream Shop   
33     North York             0.0             Clothing Store   
34     North York             0.0                Pizza Place   
47   East Toronto             0.0          Indian Restaurant   
55     North York             0.0         Italian Restaurant   
56           York             0.0     Furniture / Home Store   
59     North York             0.0                Coffee Shop   
65    Scarborough             0.0                Coffee Shop   
71    Scarborough             0.0  Middle Eastern Restaurant   
72     North York             0.0                Coffee Shop   
102     Etobicoke             0.0                 Restaurant   

    2nd Most Common Venue   3rd Most Common Venue 4th Most Common Venue  \
3             Coffee Shop              Restaurant     Accessories Store   
6                   Trail             Coffee Shop            Hobby Shop   
7             Coffee Shop     Japanese Restaurant      Asian Restaurant   
10   Fast Food Restaurant                     Gym           Pizza Place   
13            Coffee Shop     Japanese Restaurant      Asian Restaurant   
16            Pizza Place             Coffee Shop      Sushi Restaurant   
18                   Bank             Coffee Shop  Fast Food Restaurant   
23    Sporting Goods Shop  Furniture / Home Store         Grocery Store   
29      Indian Restaurant           Grocery Store               Brewery   
32   Fast Food Restaurant             Coffee Shop         Train Station   
33            Coffee Shop     Japanese Restaurant                Bakery   
34            Coffee Shop  Furniture / Home Store            Restaurant   
47            Coffee Shop    Fast Food Restaurant         Grocery Store   
55            Coffee Shop              Restaurant  Fast Food Restaurant   
56          Grocery Store          Discount Store           Video Store   
59       Ramen Restaurant             Pizza Place       Bubble Tea Shop   
65      Electronics Store  Furniture / Home Store            Restaurant   
71            Pizza Place            Burger Joint         Grocery Store   
72       Ramen Restaurant             Pizza Place       Bubble Tea Shop   
102  Gym / Fitness Center           Burrito Place           Coffee Shop   

    5th Most Common Venue  6th Most Common Venue       7th Most Common Venue  \
3            Dessert Shop  Vietnamese Restaurant         Fried Chicken Joint   
6              Restaurant            Bus Station                 Supermarket   
7            Burger Joint                    Gym                 Supermarket   
10            Gas Station                   Park                 Coffee Shop   
13           Burger Joint                    Gym                 Supermarket   
16          Grocery Store         Sandwich Place                      Bakery   
18          Grocery Store             Sports Bar           Food & Drink Shop   
23      Electronics Store                Brewery                        Bank   
29                    Gym            Supermarket                Burger Joint   
32          Grocery Store             Restaurant               Bowling Alley   
33                   Bank             Restaurant              Sandwich Place   
34                   Bank    Japanese Restaurant  Modern European Restaurant   
47                   Café             Restaurant 

 # Cluster 2 (below) neighborhoods have a lot of coffee shops but has a park too. Looks like a busy neighborhood too. Mr. Smith do not prefer this cluster. 

In [248]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
2    Downtown Toronto             1.0           Coffee Shop   
4    Downtown Toronto             1.0           Coffee Shop   
9    Downtown Toronto             1.0           Coffee Shop   
15   Downtown Toronto             1.0           Coffee Shop   
19       East Toronto             1.0                   Pub   
20   Downtown Toronto             1.0           Coffee Shop   
24   Downtown Toronto             1.0           Coffee Shop   
25   Downtown Toronto             1.0     Korean Restaurant   
30   Downtown Toronto             1.0           Coffee Shop   
31       West Toronto             1.0                  Café   
35          East York             1.0           Coffee Shop   
36   Downtown Toronto             1.0           Coffee Shop   
37       West Toronto             1.0                  Café   
41       East Toronto             1.0      Greek Restaurant   
42   Downtown Toronto             1.0           Coffee Shop   
43       West Toronto             1.0                  Café   
48   Downtown Toronto             1.0           Coffee Shop   
52         North York             1.0     Korean Restaurant   
54       East Toronto             1.0           Coffee Shop   
58        Scarborough             1.0     Convenience Store   
61    Central Toronto             1.0                  Café   
62    Central Toronto             1.0                  Bank   
67    Central Toronto             1.0           Coffee Shop   
68    Central Toronto             1.0           Coffee Shop   
69       West Toronto             1.0                  Café   
73    Central Toronto             1.0          Skating Rink   
74    Central Toronto             1.0                  Café   
75       West Toronto             1.0           Pizza Place   
76        Mississauga             1.0           Coffee Shop   
79    Central Toronto             1.0           Coffee Shop   
80   Downtown Toronto             1.0                  Café   
81       West Toronto             1.0           Coffee Shop   
83    Central Toronto             1.0                  Park   
84   Downtown Toronto             1.0                  Café   
86    Central Toronto             1.0           Coffee Shop   
88          Etobicoke             1.0                  Café   
92   Downtown Toronto             1.0           Coffee Shop   
96   Downtown Toronto             1.0                  Park   
97   Downtown Toronto             1.0           Coffee Shop   
98          Etobicoke             1.0           Coffee Shop   
99   Downtown Toronto             1.0           Coffee Shop   
100      East Toronto             1.0           Coffee Shop   

    2nd Most Common Venue          3rd Most Common Venue  \
2                 Theater                          Diner   
4                    Park               Sushi Restaurant   
9               Gastropub            Japanese Restaurant   
15                   Café                     Restaurant   
19            Coffee Shop                    Pizza Place   
20                   Café                          Hotel   
24       Ramen Restaurant                          Diner   
25                   Café                    Coffee Shop   
30                   Café                        Theater   
31                   Park                    Coffee Shop   
35                   Café               Greek Restaurant   
36                  Hotel                           Café   
37             Restaurant                            Bar   
41                   Café                    Coffee Shop   
42                   Café                          Hotel   
43            Coffee Shop                            Bar   
48                  Hotel                           Café   
52                   Café                    Pizza Place   
54                    Bar                           Café   
58               Gym Pool                Thai Restaurant   
61            College Gym                   

 # Cluster 3 (below) neighborhoods has a park but nothing much that meets Mr. Smith's preferences. It does not have a grocery store and public transport ammenities nearby

In [252]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
12   Scarborough             2.0        Breakfast Spot                  Park   
101    Etobicoke             2.0                  Park    Italian Restaurant   

           3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
12                    Playground          Burger Joint    Italian Restaurant   
101  Eastern European Restaurant        Ice Cream Shop  Gym / Fitness Center   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  ...  \
12            Yoga Studio           Event Space   Dumpling Restaurant  ...   
101  Ethiopian Restaurant            Donut Shop           Dry Cleaner  ...   

    1st Least Common Venue 2nd Least Common Venue 3rd Least Common Venue  \
12                     Zoo        Organic Grocery           Optical Shop   
101                    Zoo   Other Great Outdoors        Organic Grocery   

    4th Least Common Venue 5th Least Common Venue 6th Least Common Venue  \
12             Opera House                 Office           Noodle House   
101           Optical Shop            Opera House                 Office   

    7th Least Common Venue   8th Least Common Venue 9th Least Common Venue  \
12               Nightclub  New American Restaurant   Other Great Outdoors   
101           Noodle House                Nightclub      Other Repair Shop   

     10th Least Common Venue  
12                Nail Salon  
101  New American Restaurant  

[2 rows x 22 columns]

# Cluster 4 (below) shows neighborhood that is quite balanced. Although some neighborhoods still have the most common venue as coffee shop, is not that common compared to Cluster 1 and 2. Mr. Smith can choose from a neighborhood from this list!

In [253]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             3.0                  Park   
1         North York             3.0           Coffee Shop   
5          Etobicoke             3.0              Pharmacy   
8          East York             3.0           Pizza Place   
11         Etobicoke             3.0                  Park   
14         East York             3.0           Coffee Shop   
17         Etobicoke             3.0           Coffee Shop   
21              York             3.0    Mexican Restaurant   
22       Scarborough             3.0           Coffee Shop   
26       Scarborough             3.0                Bakery   
27        North York             3.0              Pharmacy   
28        North York             3.0                  Bank   
38       Scarborough             3.0           Coffee Shop   
39        North York             3.0         Grocery Store   
40        North York             3.0           Coffee Shop   
44       Scarborough             3.0          Intersection   
46        North York             3.0           Coffee Shop   
49        North York             3.0           Coffee Shop   
50        North York             3.0     Electronics Store   
51       Scarborough             3.0           Pizza Place   
53        North York             3.0           Coffee Shop   
57        North York             3.0     Convenience Store   
60        North York             3.0           Coffee Shop   
63              York             3.0           Pizza Place   
64              York             3.0           Coffee Shop   
66        North York             3.0                  Park   
70         Etobicoke             3.0           Pizza Place   
77         Etobicoke             3.0              Pharmacy   
78       Scarborough             3.0    Chinese Restaurant   
82       Scarborough             3.0           Coffee Shop   
85       Scarborough             3.0    Chinese Restaurant   
89         Etobicoke             3.0         Grocery Store   
90       Scarborough             3.0    Chinese Restaurant   
91  Downtown Toronto             3.0           Coffee Shop   
93         Etobicoke             3.0        Discount Store   

    2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
0       Convenience Store                    Pharmacy         Shopping Mall   
1              Boxing Gym                Hockey Arena   Sporting Goods Shop   
5           Grocery Store                Skating Rink                  Bank   
8                 Brewery        Fast Food Restaurant    Athletics & Sports   
11                  Hotel                 Pizza Place               Theater   
14                   Park                 Pizza Place        Sandwich Place   
17      Fish & Chips Shop               Shopping Mall        Shopping Plaza   
21               Pharmacy                        Park                  Bank   
22                   Park           Mobile Phone Shop    Chinese Restaurant   
26            Coffee Shop                        Bank           Gas Station   
27            Coffee Shop                        Park           Pizza Place   
28            Pizza Place           Convenience Store           Coffee Shop   
38         Discount Store          Chinese Restaurant  Fast Food Restaurant   
39                   Bank                 Gas Station   Japanese Restaurant   
40  Vietnamese Restaurant                 Pizza Place                 Hotel   
44      Convenience Store                       Diner                Bakery   
46  Vietnamese Restaurant                 Pizza Place                 Hotel   
49      Convenience Store          Athletics & Sports                Bakery   
50    Arts & Crafts Store                    Pharmacy                Bakery   
51                  Beach              Ice Cream Shop            Sports Bar   
53  Vietnamese Restaurant                 Pizza Place                 Hotel   
57           Intersection                 

# Cluster 5 (below) only has park that fits to Mr. Smith preferences. Maybe not the best!

In [254]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 5, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
45  North York             5.0                  Park                  Pool   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
45           Yoga Studio    Falafel Restaurant           Dry Cleaner   

   6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
45   Dumpling Restaurant  Eastern European Restaurant     Electronics Store   

    ... 1st Least Common Venue 2nd Least Common Venue 3rd Least Common Venue  \
45  ...                    Zoo        Organic Grocery           Optical Shop   

   4th Least Common Venue 5th Least Common Venue 6th Least Common Venue  \
45            Opera House                 Office           Noodle House   

   7th Least Common Venue   8th Least Common Venue 9th Least Common Venue  \
45              Nightclub  New American Restaurant   Other Great Outdoors   

   10th Least Common Venue  
45              Nail Salon  

[1 rows x 22 columns]

# Cluster 6 most common venue is a hotel. Nothing from the top 10 meets Mr. Smith preferences. Not this cluster!

In [256]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 6, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
94  Etobicoke             6.0                 Hotel   Rental Car Location   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
94           Coffee Shop           Yoga Studio    Falafel Restaurant   

   6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
94           Dry Cleaner   Dumpling Restaurant  Eastern European Restaurant   

    ... 1st Least Common Venue 2nd Least Common Venue 3rd Least Common Venue  \
94  ...                    Zoo   Other Great Outdoors        Organic Grocery   

   4th Least Common Venue 5th Least Common Venue 6th Least Common Venue  \
94           Optical Shop            Opera House                 Office   

   7th Least Common Venue 8th Least Common Venue 9th Least Common Venue  \
94           Noodle House              Nightclub      Other Repair Shop   

    10th Least Common Venue  
94  New American Restaurant  

[1 rows x 22 columns]

# From the analysis, it seems like neighborhoods from Cluster 4 seems like the best for Mr. Smith to settle down. He can use this cluster as a starting point to choose the neighborhood that suits him and his family the best. Summary of the neighborhoods and the location of each one is pointed in the map below. 

In [295]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
toronto_merged=toronto_merged[(toronto_merged['Cluster Labels'])==3]
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
        label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
        folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color='red',
        fill=True,
        fill_color='red',
        fill_opacity=0.7).add_to(map_clusters)
map_clusters

In [294]:
toronto_merged

Postal code           Borough  \
0          M3A        North York   
1          M4A        North York   
5          M9A         Etobicoke   
8          M4B         East York   
11         M9B         Etobicoke   
14         M4C         East York   
17         M9C         Etobicoke   
21         M6E              York   
22         M1G       Scarborough   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
38         M1K       Scarborough   
39         M2K        North York   
40         M3K        North York   
44         M1L       Scarborough   
46         M3L        North York   
49         M6L        North York   
50         M9L        North York   
51         M1M       Scarborough   
53         M3M        North York   
57         M9M        North York   
60         M3N        North York   
63         M6N              York   
64         M9N              York   
66         M2P        North York   
70         M9P         Etobicoke   
77         M9R         Etobicoke   
78         M1S       Scarborough   
82         M1T       Scarborough   
85         M1V       Scarborough   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
93         M8W         Etobicoke   

                                         Neighborhood   Latitude  Longitude  \
0                                           Parkwoods  43.753259 -79.329656   
1                                    Victoria Village  43.725882 -79.315572   
5                                    Islington Avenue  43.667856 -79.532242   
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937   
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724   
14                                   Woodbine Heights  43.695344 -79.318389   
17  Eringate, Bloordale Gardens, Old Burnhamthorpe...  43.643515 -79.577201   
21                                Caledonia-Fairbanks  43.689026 -79.453512   
22                                             Woburn  43.770992 -79.216917   
26                                          Cedarbrae  43.773136 -79.239476   
27                                  Hillcrest Village  43.803762 -79.363452   
28    Bathurst Manor, Wilson Heights, Downsview North  43.754328 -79.442259   
38        Kennedy Park, Ionview, East Birchmount Park  43.727929 -79.262029   
39                                    Bayview Village  43.786947 -79.385975   
40                                          Downsview  43.737473 -79.464763   
44                    Golden Mile, Clairlea, Oakridge  43.711112 -79.284577   
46                                          Downsview  43.739015 -79.506944   
49           North Park, Maple Leaf Park, Upwood Park  43.713756 -79.490074   
50                                      Humber Summit  43.756303 -79.565963   
51    Cliffside, Cliffcrest, Scarborough Village West  43.716316 -79.239476   
53                                          Downsview  43.728496 -79.495697   
57                                   Humberlea, Emery  43.724766 -79.532242   
60                                          Downsview  43.761631 -79.520999   
63                      Runnymede, The Junction North  43.673185 -79.487262   
64                                             Weston  43.706876 -79.518188   
66                                    York Mills West  43.752758 -79.400049   
70                                          Westmount  43.696319 -79.532242   
77  Kingsview Village, St. Phillips, Martin Grove ...  43.688905 -79.554724   
78                                          Agincourt  43.794200 -79.262029   
82            Clarks Corners, Tam O'Shanter, Sullivan  43.781638 -79.304302   
85  Milliken, Agincourt North, Steeles East, L'Amo...  43.815252 -79.284577   
89  South Steeles, Silverstone, Humbergate, Jamest...  43.739416 -79.588437   
90                      Steeles West, L'Amoreaux West  43.799525 -79.318389   
91                  